[DDP] Implement a hook which performs FunctionalSGD step.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61678

This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

rohan-varma committed:
[DDP] Implement a hook which performs FunctionalSGD step.

title

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

zhaojuanmao commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -1579,6 +1618,17 @@ def test_default_ddp_comm_hooks_nccl(self):
+    @skip_if_lt_x_gpu(2)
+    def test_hook_then_optimizer_nccl_grad_as_bucket_view(self):
+        self._test_hook_then_optimizer(gradient_as_bucket_view=True)
+
```
could we add a test when gradient_as_bucket_view and static_graph is enabled?

rohan-varma replied:
Added in the most recent version.

zhaojuanmao commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -20,6 +20,7 @@
 import torch.distributed.algorithms.ddp_comm_hooks.post_localSGD_hook as post_localSGD
 import torch.distributed.algorithms.model_averaging.averagers as averagers
 import torch.distributed.algorithms.model_averaging.utils as model_averaging_utils
+from torch.distributed.optim.functional_sgd import _FunctionalSGD
```
do we have other functional optimizers supported?

rohan-varma replied:
Not currently, but @wanchaol has implemented several of these functional optimizers for the RPC framework. It seems pretty feasible to add a `step_param` to all of those and automatically support them (such as adam, rmsprop, etc).

zhaojuanmao commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3797,6 +3798,90 @@ def test_ddp_comm_hook_logging(self):
+                lr=sgd_lr
+            )
+            allreduce_hook = default.allreduce_hook
+            opt_hook_state = default.OptimizerHookState(
+                _FunctionalSGD,
+                sgd_lr,
+            )
+            ddp_model_with_optimizer_hook.register_comm_hook(
+                None,
+                default.hook_then_optimizer(allreduce_hook, opt_hook_state),
+            )
```
Could we wrap these codes into a higher level of API for users to use? e.g. ddp_model._use_functional_optimizer( _FunctionalSGD, sgd_lr)

also, how do we prevent users to call optimizer.step() in the loop?

will allreduce_hook increase memory usage and also increase copy btw bucket tensor and tensors owned by gradBucket?

rohan-varma replied:
@zhaojuanmao Will definitely add better higher level APIs and discuss the design in follow up PRs. Would be good to land this though so we can continue to experiment (running accuracy and QPS flows now) before we support this feature officially.

> also, how do we prevent users to call optimizer.step() in the loop?
There's no way DDP can know about local optimizer AFAIK, so developer would need to know to not init an optimizer and call it in their loop before hand.

> will allreduce_hook increase memory usage and also increase copy btw bucket tensor and tensors owned by gradBucket?

The hook is sort of an interface now and in fact even when hook is not registered, we run a default allreduce hook instead of explicit call to allreduce. I believe @SciPioneer verified comm. hooks don't incur a memory penalty some time ago as well, but I'll run some tests around memory usage here as well.

zhaojuanmao reviewed (commented):
nice work! left a few comments and curious about a few questions

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3797,6 +3798,90 @@ def test_ddp_comm_hook_logging(self):
+                sgd_no_hook.step()
+                for hook_param, allreduce_param, no_hook_param in zip(
+                    ddp_model_with_optimizer_hook.parameters(),
+                    ddp_model_with_allreduce_hook.parameters(),
```
if `ddp_model_with_optimizer_hook` already gives the same results as the no-hook case, any reason we need to further compare with `ddp_model_with_allreduce_hook`?

rohan-varma replied:
@mrshenli Good point, and we already do test hook parity vs local for allreduce in separate tests (`test_ddp_hook_parity`). I'll update the test to test between model with hook + optimizer versus regular model.

mrshenli reviewed (commented):
Thanks for adding this! Do we know how much perf improvement this attains?

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma commented:
> Thanks for adding this! Do we know how much perf improvement this attains?

@mrshenli There is 1.5-3% QPS gain on internal benchmarks, will run some more to verify. After that it would be good to land so we can continue the experimentation on larger internal models where optimizer is more of a bottleneck and consider cases where optimizer does comm. (such as ZeRO, cc @andwgu)

SciPioneer commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -65,6 +65,51 @@ def decompress(fut):
+    def __init__(self, functional_optim_cls, *functional_optim_args):
+        # TODO: Support kwargs
+        self.functional_optim_cls = functional_optim_cls
+        self.functional_optim_args = functional_optim_args
```
Is it possible that we may have tensor args here? If so, shall we co-locate them on the same device?

rohan-varma replied:
@SciPioneer for SGD AFAIK the arguments are all non-tensor, but future optimizers may have tensor arguments. What does colocating the tensors on the same device (as opposed to keeping them on device user specified) allow?

SciPioneer commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -65,6 +65,51 @@ def decompress(fut):
+        hook_state, bucket: dist.GradBucket
+    ) -> torch.futures.Future:
+        # Run original hook
+        fut = hook(hook_state, bucket)
```
> will one separate CPU thread execute one hook and hold the future? is allreduce_work.wait() called inside hook() and fut value is filled inside hook() as well?

I think at this moment it is still just a future w/o completion for asynchronous. Once this hook/future appends a `.then(...)`, the successor future will wait for this work. Basically `wait()` is wrapped by `then`.

rohan-varma replied:
`wait` will be called as normal in the reducer on the future that the hook returns, and yes, the future value is filled inside the user-installed communication hook as you mentioned. I guess the hook is run by one of the autograd threads, since we call `runHook` when a bucket is ready which happens in an autograd post hook.

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3759,6 +3760,97 @@ def test_ddp_comm_hook_logging(self):
+            ]
+            # Enable determinism in cudnn operators
+            for (model, inp) in models_to_test:
+                with torch.backends.cudnn.flags(
```
Is there a global flag that we can set for all the tests in the same file? If we have more tests, we probably want to have the same setup.

rohan-varma replied:
Unfortunately simply doing the following:
```
torch.backends.cudnn.benchmark = False
torch.backends.cudnn.deterministic = True
```

ran into some assertion errors, so we weren't able to just do that. This context manger has the benefit of setting/unsetting those variables as well so they're not unintentionally set.

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -65,6 +65,51 @@ def decompress(fut):
+                    model_param,
+                    grad_tensor,
+                )
+            return [bucket.get_tensor()]
```
FYI: This should be a single tensor after #62074

rohan-varma replied:
Sounds good, I might still run a few more experiments/polishing, I'll rebase if you land that PR first

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Implement a hook which performs FunctionalSGD step."


This diff makes the following changes:
- Add `step_param` method to `_FunctionalSGD` class which is written similar to `step` but for a single param
- Implement a communication hook wrapper that runs a given comm. hook and then applies functional SGD step
- Verifies that this is equal to regular allreduce + SGD optimizer

Differential Revision: [D29701447](https://our.internmc.facebook.com/intern/diff/D29701447/)

[ghstack-poisoned]
